### PR TITLE
bug: make boto3 calls thread safe

### LINKED ISSUE #1081: Convert boto3 calls to thread safe version
boto3 is not thread safe, because why would it be?

### PR DESCRIPTION
**Pull request by jrconlin**
Closes #1081

### CODE REVIEW

File: autopush/tests/support.py [hunk 1]
```diff
@@ -44,5 +45,6 @@ def test_db(metrics=None):
         router_conf=DDBTableConfig(tablename='router'),
         router=Mock(spec=Router),
         message_conf=DDBTableConfig(tablename='message'),
-        metrics=SinkMetrics() if metrics is None else metrics
+        metrics=SinkMetrics() if metrics is None else metrics,
+        sessions=autopush.tests.sessions,
```
Comment (jrconlin) [inline autopush/tests/support.py]: for unknown reasons, nose doesn't import `sessions` from `autopush.tests` correctly (FWIW, pytest does). Thus the fully qualified path, which does resolve to the correct global pool.

File: autopush/db.py [hunk 2]
```diff
@@ -96,15 +98,30 @@
...
+SESSION_ARGS = {}
+MAX_SESSIONS = 10
+
+
+class Sessions(object):
+
+    def __init__(self, conf=SESSION_ARGS):
+        self.pool = deque(maxlen=MAX_SESSIONS)
+        session = boto3.session.Session()
+        if not conf.get("endpoint_url") and os.getenv("AWS_LOCAL_DYNAMODB"):
```
Comment (bbangert) [inline autopush/db.py]:
```
"endpoint_url" not in conf
```

File: autopush/db.py [hunk 3]
```diff
@@ -96,15 +98,30 @@
...
+        self.pool.extendleft((session.resource(
+            'dynamodb',
+            config=botocore.config.Config(
+                region_name=os.getenv("AWS_REGION_NAME", "us-east-1")
+            ),
+            **conf
+            ) for x in range(0, MAX_SESSIONS)))
+
+    def fetch(self):
+        return self.pool.pop()
```
Comment (bbangert) [inline autopush/db.py]: What happens if it runs out? Should it make another?

File: autopush/db.py [hunk 4]
```diff
@@ -96,15 +98,30 @@
...
-g_client = g_dynamodb.meta.client
+SESSION_ARGS = {}
+MAX_SESSIONS = 10
+
+
+class Sessions(object):
+
+    def __init__(self, conf=SESSION_ARGS):
+        self.pool = deque(maxlen=MAX_SESSIONS)
+        session = boto3.session.Session()
```
Comment (bbangert) [inline autopush/db.py]: boto_session might avoid pain in the future as we track various 'session' concepts around.

File: autopush/db.py [hunk 5]
```diff
@@ -146,20 +163,21 @@ def make_rotating_tablename(prefix, delta=0, date=None):
 
 def create_rotating_message_table(prefix="message", delta=0, date=None,
                                   read_throughput=5,
-                                  write_throughput=5):
-    # type: (str, int, Optional[datetime.date], int, int) -> Table
+                                  write_throughput=5,
+                                  session=None):
```
Comment (bbangert) [inline autopush/db.py]: Above we have 'session' which is a boto session.Session, yet here its a session.resource which is not the same. Here it's actually a DynamoDB resource created with the boto session.Session.

**Comment by bbangert** [comment]
Why not make a thread-local module global we initialize and use (sorta like the boto2 one) instead of passing it all over?

File: autopush/db.py (hunk 2 above)
Comment (jrconlin) [reply autopush/db.py]: It normally shouldn't be. If it's not specified, boto3 will use it's own, else it will use whatever you note. (Passing `None` is problematic) We do pass an `endpoint_url` as part of testing.

File: autopush/db.py [hunk 6]
```diff
@@ -96,15 +98,30 @@
...
+
+
+class Sessions(object):
+
+    def __init__(self, conf=SESSION_ARGS):
+        self.pool = deque(maxlen=MAX_SESSIONS)
+        session = boto3.session.Session()
+        if not conf.get("endpoint_url") and os.getenv("AWS_LOCAL_DYNAMODB"):
+            conf["endpoint_url"] = os.getenv("AWS_LOCAL_DYNAMODB")
+        self.pool.extendleft((session.resource(
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply autopush/db.py]: Heh, copied what we do for apns2. (and @pjenvey usually yells at me to use list comprehension)

I'm fine either way.

**Comment by jrconlin** [comment]
we tend to make a number of thread calls when we defer for the database. the boto_session is tied to an instance rather than a thread where possible to ensure that meta information remains as consistent as possible.

File: autopush/db.py [hunk 7]
```diff
@@ -650,65 +741,97 @@ def fetch_timestamp_messages(
...
                         :class:`autopush.metrics.IMetrics` interface.
 
         """
-        self.table = table
+        self.conf = conf
         self.metrics = metrics
         self._max_ttl = max_ttl
+        self._session = None
+        self._cached_table = None
+        self._sessions = boto_sessions or BotoSessions()
```
Comment (pjenvey) [inline autopush/db.py]: I think boto_sessions is always specified?

File: autopush/main.py [hunk 8]
```diff
@@ -148,7 +148,8 @@ def main(cls, args=None, use_files=True):
             firehose_delivery_stream=ns.firehose_stream_name
         )
         try:
-            app = cls.from_argparse(ns)
+            cls.argparse = cls.from_argparse(ns, sessions=sessions)
```
Comment (pjenvey) [inline autopush/main.py]: is this assignment needed?

File: tox.ini [hunk 9]
```diff
@@ -1,18 +1,19 @@
 [tox]
-envlist = py27,pypy,flake8,py36-mypy
+#envlist = py27,pypy,flake8,py36-mypy
+envlist = pypy,flake8
```
Comment (pjenvey) [inline tox.ini]: !

File: autopush/webpush_server.py [hunk 10]
```diff
@@ -630,9 +677,14 @@ def process(self,
         if not valid:
             return UnregisterErrorResponse(error_msg=msg)
 
-        message = self.db.message_tables[command.message_month]
+        session = self.db.sessions.fetch()
```
Comment (pjenvey) [inline autopush/webpush_server.py]: session unused here

File: autopush/webpush_server.py [hunk 11]
```diff
@@ -493,46 +516,66 @@ class DeleteMessageCommand(ProcessorCommand):
...
+                          boto_sessions=self.db.sessions)
         message.delete_message(notif)
         return DeleteMessageResponse()
 
 
 class DropUserCommand(ProcessorCommand):
     def process(self, command):
         # type: (DropUser) -> DropUserResponse
-        self.db.router.drop_user(command.uaid.hex)
+        session = self.db.sessions.fetch()
```
Comment (pjenvey) [inline autopush/webpush_server.py]: unused

File: autopush/main.py [hunk 12]
```diff
@@ -63,11 +63,11 @@ class AutopushMultiService(MultiService):
 
     THREAD_POOL_SIZE = 50
 
-    def __init__(self, conf):
+    def __init__(self, conf, sessions=None):
```
Comment (pjenvey) [inline autopush/main.py]: nitpicky: ideally I think DatabaseManager being more responsible for creating the BotoSessions (maybe taking a config for it that includes the endpoint_url, say for tests) would avoid the hooks here. A lot like it is responsible for the Table objects, but also allowing overrides of those objects for tests.

Unfortunately our tests around main'ish functions are currently kind of crappy and would probably require some annoying monkey patches to force the right test configuration

**Comment by pjenvey** [comment]
so we definitely need to rename 'session' being that it's a resource. BotoSessions doesn't seem appropriate either. My other main issue with the naming in the current patch is "boto_session" vs "boto_sessions" seemed like a typo waiting to happen

**Comment by pjenvey** [comment]
also not necessary for now but the 'session' fetch/release screams for a context manager. I imagine one that could optionally handle being specified an existing session and handling the 'release = False' noop behavior for the caller

File: tox.ini (hunk 9 above)
Comment (jrconlin) [reply tox.ini]: crap. sorry, that's because I was trying to force test against pypy.

File: autopush/db.py [hunk 13]
```diff
@@ -96,15 +98,42 @@
...
-# additional config options
-g_dynamodb = boto3.resource(
-    'dynamodb',
-    config=botocore.config.Config(
-        region_name=os.getenv("AWS_REGION_NAME", "us-east-1")
-    )
-)
-g_client = g_dynamodb.meta.client
+SESSION_ARGS = {}
+MAX_SESSIONS = 10
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply autopush/db.py]: Oh, good call!

**Comment by jrconlin** [comment]
I agree about the context manager. I mostly just wanted to get this landed and skipping the context manager made it easier for me to follow what was going on. It also let me control very tightly exactly when we fetch and release a resource handle, which was useful during debugging.

File: autopush/webpush_server.py [hunk 14]
```diff
@@ -360,7 +362,12 @@ def process(self, hello):
 
         # Save the UAID as register_user removes it
         uaid = user_item["uaid"]  # type: str
-        success, _ = self.db.router.register_user(user_item)
+        session = self.db.resources.fetch()
```
Comment (bbangert) [inline autopush/webpush_server.py]: session -> ddb_resource

File: autopush/websocket.py [hunk 15]
```diff
@@ -1047,17 +1086,23 @@ def _monthly_transition(self):
 
         """
         # Get the current channels for this month
-        _, channels = self.ps.message.all_channels(self.ps.uaid)
+        message = self.db.message_table(self.ps.message_month)
+        _, channels = message.all_channels(self.ps.uaid)
```
Comment (bbangert) [inline autopush/websocket.py]: How come these commands don't need a ddb_resource?

Comment (jrconlin) [reply autopush/websocket.py]: `self.db.message_table` automatically uses the resource pool already assigned in `db`. Only stand alone functions require a specific resource handle to be passed in.

File: autopush/websocket.py [hunk 16]
```diff
@@ -249,7 +249,7 @@ def __attrs_post_init__(self):
     def message(self):
         # type: () -> Message
         """Property to access the currently used message table"""
-        return self.db.message_tables[self.message_month]
+        return Message(self.message_month, resource_pool=self.db.resources)
```
Comment (jrconlin) [inline autopush/websocket.py]: No idea why coverage doesn't see this. Maybe because it's a property?

File: autopush/db.py [hunk 17]
```diff
@@ -293,28 +342,30 @@ def create_router_table(tablename="router", read_throughput=5,
...
 
 
 def _expiry(ttl):
     return int(time.time() + ttl)
 
 
 def get_router_table(tablename="router", read_throughput=5,
-                     write_throughput=5):
+                     write_throughput=5, boto_resource=None):
     # type: (str, int, int) -> Table
```
Comment (bbangert) [inline autopush/db.py]: Type sig change.

File: autopush/db.py [hunk 18]
```diff
@@ -337,7 +389,7 @@ def preflight_check(message, router, uaid="deadbeef00000000deadbeef00000000"):
     # Verify tables are ready for use if they just got created
     ready = False
     while not ready:
-        tbl_status = [x.table_status() for x in [message, router]]
+        tbl_status = [x.table_status()for x in [message, router]]
```
Comment (bbangert) [inline autopush/db.py]: Lost the space after status().

File: autopush/db.py [hunk 19]
```diff
@@ -694,24 +748,37 @@ def update_last_message_read(self, uaid, timestamp):
...
                         :class:`autopush.metrics.IMetrics` interface.
+        :param resource_pool: Pool of :class:`autopush.db.BotoResources`
 
         """
-        self.table = table
+        self.conf = conf
         self.metrics = metrics
         self._max_ttl = max_ttl
+        self._cached_table = None
+        self._resources = resource_pool or BotoResources()
```
Comment (bbangert) [inline autopush/db.py]: resource_pool is required, so this shouldn't happen?

File: autopush/db.py [hunk 20]
```diff
@@ -694,24 +748,37 @@ def update_last_message_read(self, uaid, timestamp):
...
+        self._resource = None
```
Comment (bbangert) [inline autopush/db.py]: What is this for?

File: autopush/db.py [hunk 21]
```diff
@@ -694,24 +748,37 @@ def update_last_message_read(self, uaid, timestamp):
...
+                table = get_router_table(boto_resource=boto_resource)
+        return table
 
     def table_status(self):
-        return self.table.table_status
+        return self.table().table_status
 
     def get_uaid(self, uaid):
-        # type: (str) -> Item
+        # type: (str, ServiceResource) -> Item
```
Comment (bbangert) [inline autopush/db.py]: Type sig mismatch.

File: autopush/db.py [hunk 22]
```diff
@@ -851,6 +920,7 @@ def drop_old_users(self, months_ago=2):
             quickly as possible.
 
         :param months_ago: how many months ago since the last connect
+        :param boto_resource: ServiceResource for thread
```
Comment (bbangert) [inline autopush/db.py]: Param doesn't exist here.

File: autopush/db.py [hunk 23]
```diff
@@ -96,16 +100,58 @@
...
+            del(conf["endpoint_url"])
+        self.conf = conf
+        self.ddb_data = threading.local()
+
+    def __enter__(self):
+        self._resource = self.fetch()
+        return self._resource
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.release(self._resource)
```
Comment (bbangert) [inline autopush/db.py]: Why is this necessary?

File: autopush/db.py [hunk 24]
```diff
@@ -293,37 +297,47 @@ def create_router_table(tablename="router", read_throughput=5,
...
+        boto_resource      # type: DynamoDBResource
+        ):
+    # type (...) -> DynamoDBTable
     """Private common function to make a table with a table func"""
-    if not table_exists(tablename):
-        return table_func(tablename, read_throughput, write_throughput)
+    if not boto_resource:
+        raise AutopushException("No boto3 resource provided for _make_table")
+    if not table_exists(tablename, boto_resource):
+        return table_func(tablename, read_throughput, write_throughput,
```
Comment (pjenvey) [inline autopush/db.py]: This should return the table wrapped in DynamoDBTable like the clause below

File: autopush/db.py [hunk 25]
```diff
@@ -432,45 +445,70 @@ def generate_last_connect_values(date):
...
+        :param boto_resource: DynamoDBResource for thread
 
         """
-        self.table = table
-        self.metrics = metrics
+        self.tablename = tablename
         self._max_ttl = max_ttl
+        self.resource = boto_resource
+
+    def table(self, tablename=None):
```
Comment (pjenvey) [inline autopush/db.py]: I don't think the table() method and all the associated changes for it are necessary now that the underlying 'table' object here would be the thread local wrapper

File: autopush/db.py [hunk 26]
```diff
@@ -432,45 +445,70 @@ def generate_last_connect_values(date):
...
 
-def table_exists(tablename, client=None):
-    """Determine if the specified Table exists"""
-    if not client:
-        client = g_client
-    return tablename in list_tables(client)
+class DynamoDBResource(threading.local):
+    def __init__(self, **kwargs):
+        conf = kwargs
+        if not conf.get("endpoint_url"):
```
Comment (pjenvey) [inline autopush/db.py]: this is now aws_ddb_endpoint. we don't need to handle AWS_LOCAL_DYNAMODB env here any longer since configargparse is now doing it for us

File: autopush/db.py [hunk 27]
```diff
@@ -432,45 +445,70 @@ def generate_last_connect_values(date):
...
+            conf["endpoint_url"] = os.getenv("AWS_LOCAL_DYNAMODB")
+        # If there is no endpoint URL, we must delete the entry
+        if "endpoint_url" in conf and not conf["endpoint_url"]:
+            del(conf["endpoint_url"])
+        region = conf.get("region_name",
+                          os.getenv("AWS_REGION_NAME", "us-east-1"))
+        if "region_name" in conf:
+            del(conf["region_name"])
+        self.conf = conf
```
Comment (pjenvey) [inline autopush/db.py]: we don't need to keep self.conf here

File: autopush/tests/__init__.py [hunk 28]
```diff
@@ -4,48 +4,45 @@
...
     # Setup the necessary message tables
     message_table = os.environ.get("MESSAGE_TABLE", "message_int_test")
-
-    create_rotating_message_table(prefix=message_table, delta=-1)
-    create_rotating_message_table(prefix=message_table)
+    create_rotating_message_table(prefix=message_table, delta=-1,
+                                  boto_resource=boto_resource)
+    create_rotating_message_table(prefix=message_table,
+                                  boto_resource=boto_resource)
+    pool = reactor.getThreadPool()
```
Comment (pjenvey) [inline autopush/tests/__init__.py]: we don't need the pool adjustment size any longer

File: autopush/db.py (hunk 24 above)
Comment (jrconlin) [reply autopush/db.py]: Sorry, I'm not sure I follow.

If a table doesn't exist, it must be created. We're passing the `create_router_table` function as the `table_func` param, since it contains the full table schema, which would not be available to the generic `DynamoDBTable`.

Are you suggesting that I move `create_router_table` as a class method of `DynamoDBTable` (which I could understand), but then, shouldn't I also do the same for `create_initial_message_tables`?

File: autopush/db.py (hunk 25 above)
Comment (jrconlin) [reply autopush/db.py]: We use `table()` to select between Message tables, so we'd still need to call DynamoDBTable() to fetch the correct reference to the table we'd want. 

For `Router` it makes more sense to drop `table()`, since there's only going to be one.

File: autopush/db.py (hunk 24 above)
Comment (pjenvey) [reply autopush/db.py]: What I mean is I think get_router_table and get_rotating_message_table should both return a thread safe DynamoDBTable wrapper. get_router_table/_make_table as is are returning either a DynamoDBTable wrapper if the table exists, otherwise the low level boto3 Table. That doesn't make much sense

File: autopush/db.py (hunk 25 above)
Comment (pjenvey) [reply autopush/db.py]: What I mean is originally we had autopush.db.Message/Router with a thread safe '.table' boto2 Table object.

If we're turning '.table' into a thread safe DynamoDBTable wrapper, doesn't the need for a table() method (and the change to db.message_tables) go away? All this can continue working as before as long as the .table is thread safe (unless I'm missing something)

File: autopush/db.py (hunk 24 above)
Comment (jrconlin) [reply autopush/db.py]: Ah! Ok. I agree.

Comment (jrconlin) [reply autopush/db.py]: Modified `_create_*_table` to return `DynamoDBTable` instances.

File: autopush/db.py [hunk 29]
```diff
@@ -432,42 +444,60 @@ def generate_last_connect_values(date):
...
-            yield table
-        start_table = result.get('LastEvaluatedTableName', None)
-        if not start_table:
-            break
+def table_exists(tablename, boto_resource=None):
+    # type: (str, DynamoDBResource) -> bool
+    """Determine if the specified Table exists"""
+    try:
+        return DynamoDBTable(boto_resource, tablename).table_status in [
+            'CREATING', 'UPDATING', 'ACTIVE']
```
Comment (bbangert) [inline autopush/db.py]: Might want to use the Waiter thing to wait on creating/updating to finish, as other things may try to use the table immediately after this returns, which would fail if its not ACTIVE. That would only happen on startup I think, but could be problematic. Having a separate PR is fine.

File: autopush/db.py [hunk 30]
```diff
@@ -432,42 +444,60 @@ def generate_last_connect_values(date):
...
+        return getattr(self._table, name)
 
 
 class Message(object):
     """Create a Message table abstraction on top of a DynamoDB Table object"""
-    def __init__(self, table, metrics, max_ttl=MAX_EXPIRY):
-        # type: (Table, IMetrics) -> None
+    def __init__(self, table, metrics=None, boto_resource=None,
+                 max_ttl=MAX_EXPIRY):
+        # type: (DynamoDBTable, IMetrics, DynamoDBResource, int) -> None
```
Comment (bbangert) [inline autopush/db.py]: Why does it need the boto_resource?

File: autopush/web/health.py [hunk 31]
```diff
@@ -58,12 +59,12 @@ def _check_error(self, failure, name):
...
         cause = self._health_checks[name] = {"status": "NOT OK"}
-        if failure.check(InternalServerError):
+        if failure.check(InternalServerError):      # pragma nocover
             cause["error"] = "Server error"
-        elif failure.check(MissingTableException):
+        elif failure.check(MissingTableException):  # pragma nocover
             cause["error"] = failure.getErrorMessage()
         else:
-            cause["error"] = "Internal error"
+            cause["error"] = "Internal error"       # pragma nocover
```
Comment (bbangert) [inline autopush/web/health.py]: We had tests that covered these, now we don't?

File: autopush/db.py (hunk 29 above)
Comment (jrconlin) [reply autopush/db.py]: I'm not sure. `table_exists` is only called before a table creation command. In those situations. The command returns that the table either exists or is in the process of creation. 

`preflight_check()` already has a table ready wait routine, and is part of the startup for the process. 

Do you believe another delay would be useful?

File: autopush/db.py (hunk 30 above)
Comment (jrconlin) [reply autopush/db.py]: mostly for consistency and testing purposes. `DynamoDBTable` still needs to pull from the resource, so having the explicit resource passed ensures that nothing "magical" happens.

File: autopush/web/health.py (hunk 31 above)
Comment (jrconlin) [reply autopush/web/health.py]: Unfortunately, the test that originally triggered that had to be dropped, and getting the code to exhibit the behavior was far more difficult than expected, mostly due to the `thread.local` element isolating portions of the code from each other. I should revisit this code and see if I can enforce this behavior, but for now, I felt it's ok to have this as a "catch" behavior for edge cases.

Comment (jrconlin) [reply autopush/web/health.py]: Added #1114

File: autopush/db.py [hunk 32]
```diff
@@ -145,22 +139,25 @@ def make_rotating_tablename(prefix, delta=0, date=None):
...
     try:
-        table = g_dynamodb.Table(tablename)
-        if table.table_status == 'ACTIVE':  # pragma nocover
+        table = DynamoDBTable(boto_resource, tablename)
+        if table.table_status == 'ACTIVE':
             return table
     except ClientError as ex:
         if ex.response['Error']['Code'] != 'ResourceNotFoundException':
-            # If we hit this, our boto3 is misconfigured and we need to bail.
-            raise ex  # pragma nocover
```
Comment (pjenvey) [inline autopush/db.py]: Might we still want to raise exceptions here (I'm not sure)? Either way this block of code isn't necessary any longer as it's a "pass" for any exception

File: autopush/db.py [hunk 33]
```diff
@@ -694,18 +724,25 @@ def update_last_message_read(self, uaid, timestamp):
 
 class Router(object):
     """Create a Router table abstraction on top of a DynamoDB Table object"""
-    def __init__(self, table, metrics, max_ttl=MAX_EXPIRY):
-        # type: (Table, IMetrics, int) -> None
+    def __init__(self, conf, metrics, max_ttl=MAX_EXPIRY, resource=None):
```
Comment (pjenvey) [inline autopush/db.py]: FYI I don't we need this change from table to conf any longer nor the _cached_table var below. Changing the conf argument back will require a number of changes to test_db so I'm not going to cry over this

File: autopush/db.py (hunk 32 above)
Comment (jrconlin) [reply autopush/db.py]: ah! Good catch, thanks!

**Comment by jrconlin** [comment]
Added merge fixes for `test_rs_integration.py`

File: autopush/tests/test_rs_integration.py [hunk 34]
```diff
@@ -104,14 +104,19 @@ def setUp(self):
         )
 
         # Endpoint HTTP router
-        self.ep = ep = EndpointApplication(ep_conf)
-        ep.db.client = db.g_client
+        self.ep = ep = EndpointApplication(
+            ep_conf,
+            resource=autopush.tests.boto_resource
+        )
```
Comment (jrconlin) [inline autopush/tests/test_rs_integration.py]: Merge change to use boto_resource for testing.
